dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/Edge.java

```diff
@@ -31,34 +56,38 @@ public String toString() {
     public Tuple toTuple() {
         return new Tuple(first.getEvent(), second.getEvent());
     }
+    public Edge inverse() { return new Edge(second, first, time, derivLength); }
 
-    public Edge getInverse() { return new Edge(second, first); }
 
-    public boolean isCrossEdge() {
-        return !first.getThread().equals(second.getThread());
+    public boolean isInternal() { return first.getThread() == second.getThread();}
+    public boolean isExternal() {
+        return !isInternal();
     }
 
     public boolean isBackwardEdge() {
-        return !isCrossEdge() && first.getLocalId() > second.getLocalId();
+        return isInternal() && first.getLocalId() > second.getLocalId();
     }
-
     public boolean isForwardEdge() {
-        return !isCrossEdge() && first.getLocalId() < second.getLocalId();
+        return isInternal() && first.getLocalId() < second.getLocalId();
     }
-
     public boolean isLoop() {
         return first.equals(second);
     }
-
     public boolean isLocEdge()  { return first.isMemoryEvent() && second.isMemoryEvent()
-            && first.getAccessedAddress().compareTo(second.getAccessedAddress()) == 0; }
+            && first.getAccessedAddress().equals(second.getAccessedAddress()); }
 
+    public Edge with(Timestamp time) { return new Edge(first, second, time, derivLength); }
+    public Edge with(int derivLength) { return new Edge(first, second, time, derivLength); }
+    public Edge with(Timestamp time, int derivLength) { return new Edge(first, second, time, derivLength); }
```

Comment:
each of these creates a new instance of `Edge` and leaves the old one dangling. Wouldn't be better to do
```
public Edge with(Timestamp time) {
  this.time = time;
  return this;
}
```
to avoid this?

Reply:
They are intended to not change the current instance. ```Edge``` is an immutable class and the ```with``` functions are a common idiom to get new instances that are only slightly different to the original edge.